### ISSUE TITLE
Concern regarding parameter initialization in PoincareLinear module

### ISSUE THREAD

**Issue by nbuton** [mil-tokyo/hyperbolic_nn_plusplus#2]
Dear Authors,

Thank you for your work on this project. I have a concern regarding the initialization of parameters in the [PoincareLinear module](https://github.com/mil-tokyo/hyperbolic_nn_plusplus/blob/28737e22822562ac18d5ea03f8c3e3929e945a83/geoopt_plusplus/modules/linear.py#L29), specifically with regard to the two parameters defined: one for the weight and the other for the norm of the weight.

It appears that the two parameters are defined independently, which could lead to inconsistencies when the optimizer changes the weight and weight norm separately. I believe that these two parameters should be connected in order to ensure that they are changed together. More precisely, only weight_v should be a parameter and weigh_g should be computed in the forward pass.

In order to validate my hypothesis, I recommend temporarily adding an assert statement in the forward pass as shown below:
```
assert torch.sum(torch.abs(self.weight_v.norm(dim=0)-self.weight_g))<0.1
```
This will not verify the assertion very quickly.

I could be wrong, but I would appreciate your clarification on this matter.

Thank you in advance.

**Comment by abfariah** [comment]
I very much agree with @nbuton, as I also was astonished by this parameter initialization. It would be great to hear from the authors.

**Comment by nhamanasu** [comment]
Dear @nbuton and @abfariah,

First, I sincerely apologize for the delayed response. As the first author of the paper, I would like to address your concerns regarding the parameter initialization in the PoincareLinear module.

The parameter initialization approach we implemented follows a similar concept to [weight normalization](https://pytorch.org/docs/stable/generated/torch.nn.utils.weight_norm.html) in conventional deep learning. While weight_g and weight_v are initially calculated to represent the norm and direction respectively during initialization, they are intentionally designed to be independent parameters during the training process. This means that weight_v's norm is not required to match weight_g after initialization.

This design choice directly corresponds to the implementation in [unidirectional_poincare_mlr](https://github.com/mil-tokyo/hyperbolic_nn_plusplus/blob/main/geoopt_plusplus/modules/multinomial_logistic_regression.py#L64), which explicitly takes the norm and direction of the weight tensor as separate arguments. This aligns with our treatment of z_k's norm and direction in Eq. (6) of [our paper](https://arxiv.org/abs/2006.08210).

While it would be possible to define a single parameter (like self.weight) and decompose it into norm and direction during each forward pass, in those days when I wrote the paper, the experiments showed that this approach led to training instability within our experimental scope. This observation led us to choose the current parameterization with separate parameters.

Thank you for bringing up this important point for discussion. I hope this clarifies our design decisions and implementation choices.

Best regards,
